src/main/java/com/connexta/search/index/IndexServiceImpl.java

```diff
@@ -0,0 +1,124 @@
+    } catch (final Exception e) {
+      throw new IndexException(
+          INTERNAL_SERVER_ERROR,
+          String.format("Unable to query index for datasetId=%s", datasetId),
+          e);
+    }
+    if (idAlreadyExists) {
+      throw new IndexException(
+          BAD_REQUEST,
+          String.format("datasetId=%s already exists. Overwriting is not supported", datasetId));
+    }
+
+    final Resource resource;
+    try {
+      final HttpStatus expectedHttpStatus = HttpStatus.OK;
+      resource =
+          webClient
+              .get()
+              .uri(irmUri)
+              .retrieve()
+              .onStatus(
+                  httpStatus -> !expectedHttpStatus.equals(httpStatus),
+                  clientResponse ->
+                      Mono.error(
+                          () ->
+                              new IndexException(
+                                  BAD_REQUEST,
+                                  String.format(
+                                      "Excepted %s but received status code of %s when getting resource at irmUri=%s",
+                                      expectedHttpStatus, clientResponse.statusCode(), irmUri))))
+              .bodyToMono(Resource.class)
+              .block();
+    } catch (final Exception e) {
+      throw new IndexException(
+          BAD_REQUEST, String.format("Unable to complete GET request to irmUri=%s", irmUri), e);
+    }
+
+    if (resource == null) {
+      throw new IndexException(
+          BAD_REQUEST, String.format("Unable to complete GET request to irmUri=%s", irmUri));
+    }
+
+    final String contents;
+    try (final InputStream inputStream = resource.getInputStream()) {
+      try {
+        // TODO Parse IRM instead of JSON CST
+        contents = getElement(parseJson(inputStream), EXT_EXTRACTED_TEXT);
+      } catch (IOException e) {
+        throw new IndexException(
+            INTERNAL_SERVER_ERROR,
+            String.format("Unable to convert InputStream to JSON for irmUri=%s", irmUri),
+            e);
+      }
+    } catch (IOException e) {
+      throw new IndexException(
+          BAD_REQUEST, String.format("Unable to get InputStream from irmUri=%s", irmUri), e);
+    }
+
+    log.info("Attempting to index datasetId={}", datasetId);
+    try {
+      crudRepository.save(new Index(datasetId, contents, irmUri.toString()));
+    } catch (final Exception e) {
+      throw new IndexException(INTERNAL_SERVER_ERROR, "Unable to save index", e);
+    }
+    log.info("Successfully indexed datasetId={}", datasetId);
+  }
+
+  private static JsonNode parseJson(InputStream stream) throws IOException {
+    final ObjectMapper objectMapper;
+    objectMapper = new ObjectMapper();
+    return objectMapper.readTree(stream);
+  }
+
+  private static String getElement(JsonNode json, String fieldName) throws IndexException {
+    if (json.get(fieldName) == null) {
+      throw new IndexException(INTERNAL_SERVER_ERROR, "JSON is malformed");
```

Comment:
Maybe more appropriate to say that no value for the fieldName is present in the json.